### PR TITLE
Add metering api

### PR DESCRIPTION
**Pull request by zheng1**
**What type of PR is this?**
/hold
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for reviewers**:
```
```

**Additional documentation, usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kapis/metering/v1alpha1/register.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+		Param(ws.QueryParameter("end", "End time of query. Use **start** and **end** to retrieve metric data over a time span. It is a string with Unix time format, eg. 1561939200. ").DataType("string").Required(false)).
+		Param(ws.QueryParameter("step", "Time interval. Retrieve metric data at a fixed interval within the time range of start and end. It requires both **start** and **end** are provided. The format is [0-9]+[smhdwy]. Defaults to 10m (i.e. 10 min).").DataType("string").DefaultValue("10m").Required(false)).
+		Param(ws.QueryParameter("time", "A timestamp in Unix time format. Retrieve metric data at a single point in time. Defaults to now. Time and the combination of start, end, step are mutually exclusive.").DataType("string").Required(false)).
+		Metadata(restfulspec.KeyOpenAPITags, []string{constants.PodMetersTag}).
+		Writes(model.Metrics{}).
+		Returns(http.StatusOK, respOK, model.Metrics{})).
+		Produces(restful.MIME_JSON)
+
+	ws.Route(ws.GET("/namespaces/{namespace}/workloads/{kind}/{workload}/pods").
```
Comment (wansir) [inline pkg/kapis/metering/v1alpha1/register.go:265]: Suggested API definition `/namespaces/{namespace}/{workloads}/{workload}/pods`.

File: pkg/kapis/metering/v1alpha1/register.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+		Param(ws.QueryParameter("sort_metric", "Sort pods by the specified metric. Not applicable if **start** and **end** are provided.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("sort_type", "Sort order. One of asc, desc.").DefaultValue("desc.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("page", "The page number. This field paginates result data of each metric, then returns a specific page. For example, setting **page** to 2 returns the second page. It only applies to sorted metric data.").DataType("integer").Required(false)).
+		Param(ws.QueryParameter("limit", "Page size, the maximum number of results in a single page. Defaults to 5.").DataType("integer").Required(false).DefaultValue("5")).
+		Metadata(restfulspec.KeyOpenAPITags, []string{constants.PodMetersTag}).
+		Writes(model.Metrics{}).
+		Returns(http.StatusOK, respOK, model.Metrics{})).
+		Produces(restful.MIME_JSON)
+
+	ws.Route(ws.GET("/nodes/{node}/pods/{pod}").
```
Comment (wansir) [inline pkg/kapis/metering/v1alpha1/register.go:305]: Suggested API definition `/pods/{pod}`, node name in API path is unnecessary.

File: pkg/kapis/metering/v1alpha1/register.go (hunk 1 above)
Comment (yunkunrao) [reply pkg/kapis/metering/v1alpha1/register.go:265]: got it

File: pkg/kapis/metering/v1alpha1/register.go (hunk 2 above)
Comment (yunkunrao) [reply pkg/kapis/metering/v1alpha1/register.go:305]: got it

**Comment by zheng1** [comment]
@yunkunrao Please fix the build error

**Comment by zheng1** [comment]
@yunkunrao Please resolve those conflict files

File: pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -61,6 +61,7 @@ import (
 	devopsv1alpha2 "kubesphere.io/kubesphere/pkg/kapis/devops/v1alpha2"
 	devopsv1alpha3 "kubesphere.io/kubesphere/pkg/kapis/devops/v1alpha3"
 	iamapi "kubesphere.io/kubesphere/pkg/kapis/iam/v1alpha2"
+	meteringv1alpha3 "kubesphere.io/kubesphere/pkg/kapis/metering/v1alpha1"
```
[the remark this answers is not in the document]
Comment (yunkunrao) [reply pkg/apiserver/apiserver.go:64]: got it

File: pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -215,12 +216,13 @@ func (s *APIServer) installKubeSphereAPIs() {
 	urlruntime.Must(configv1alpha2.AddToContainer(s.container, s.Config))
 	urlruntime.Must(resourcev1alpha3.AddToContainer(s.container, s.InformerFactory, s.RuntimeCache))
 	urlruntime.Must(monitoringv1alpha3.AddToContainer(s.container, s.KubernetesClient.Kubernetes(), s.MonitoringClient, s.MetricsClient, s.InformerFactory, s.OpenpitrixClient))
+	urlruntime.Must(meteringv1alpha3.AddToContainer(s.container, s.KubernetesClient.Kubernetes(), s.MonitoringClient, s.InformerFactory, s.OpenpitrixClient, s.RuntimeCache))
```
[the remark this answers is not in the document]
Comment (yunkunrao) [reply pkg/apiserver/apiserver.go:219]: got it

**Comment by benjaminhuo** [comment]
Basically, This PR looks good to me.
@junotx Pls help to take a look at this PR, especially if it breaks existing monitoring  API

File: pkg/kapis/tenant/v1alpha2/register.go [hunk 5]
```diff
@@ -288,6 +297,50 @@ func AddToContainer(c *restful.Container, factory informers.InformerFactory, k8s
 		Writes(auditingv1alpha1.APIResponse{}).
 		Returns(http.StatusOK, api.StatusOK, auditingv1alpha1.APIResponse{}))
 
+	ws.Route(ws.GET("/meterings").
```
Comment (zryfish) [inline pkg/kapis/tenant/v1alpha2/register.go:300]: Does `metering` have a plural?

File: pkg/kapis/tenant/v1alpha2/register.go [hunk 6]
```diff
@@ -288,6 +297,50 @@ func AddToContainer(c *restful.Container, factory informers.InformerFactory, k8s
...
+		Param(ws.QueryParameter("sort_type", "Sort order. One of asc, desc.").DefaultValue("desc.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("page", "The page number. This field paginates result data of each metric, then returns a specific page. For example, setting **page** to 2 returns the second page. It only applies to sorted metric data.").DataType("integer").Required(false)).
+		Param(ws.QueryParameter("limit", "Page size, the maximum number of results in a single page. Defaults to 5.").DataType("integer").Required(false).DefaultValue("5")).
+		Param(ws.QueryParameter("storageclass", "The name of the storageclass.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("pvc_filter", "The PVC filter consists of a regexp pattern. It specifies which PVC data to return.").DataType("string").Required(false)).
+		Metadata(restfulspec.KeyOpenAPITags, []string{constants.WorkspaceMetersTag}).
+		Writes(monitoring.Metrics{}).
+		Returns(http.StatusOK, api.StatusOK, monitoring.Metrics{}))
+
+	ws.Route(ws.GET("/namespaces/{namespace}/metering_hierarchy").
```
Comment (zryfish) [inline pkg/kapis/tenant/v1alpha2/register.go:329]: how about `/namespaces/{namespace}/metering/hierarchy`

File: pkg/kapis/tenant/v1alpha2/register.go [hunk 7]
```diff
@@ -288,6 +297,50 @@ func AddToContainer(c *restful.Container, factory informers.InformerFactory, k8s
...
+		To(handler.QueryMeteringsHierarchy).
+		Param(ws.PathParameter("namespace", "Namespace name.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("metrics_filter", "The metric name filter consists of a regexp pattern. It specifies which metric data to return. For example, the following filter matches both workspace CPU usage and memory usage: `meter_pod_cpu_usage|meter_pod_memory_usage_wo_cache`.").DataType("string").Required(false)).
+		Param(ws.QueryParameter("time", "A timestamp in Unix time format. Retrieve metric data at a single point in time. Defaults to now. Time and the combination of start, end, step are mutually exclusive.").DataType("string").Required(false)).
+		Doc("get current metering hierarchies info in last one hour").
+		Writes(metering.ResourceStatistic{}).
+		Returns(http.StatusOK, api.StatusOK, metering.ResourceStatistic{}))
+
+	ws.Route(ws.GET("/metering/price_info").
```
Comment (zryfish) [inline pkg/kapis/tenant/v1alpha2/register.go:338]: `metering/price` would be enough

**Comment by zryfish** [comment]
Overall, this PR is ok, but there are minor issues that need to be fixed.  Please solve the conflicts and add more comments

**Comment by zryfish** [comment]
@yunkunrao Please take a look

File: pkg/kapis/tenant/v1alpha2/register.go (hunk 5 above)
Comment (yunkunrao) [reply pkg/kapis/tenant/v1alpha2/register.go:300]: got it

File: pkg/kapis/tenant/v1alpha2/register.go (hunk 6 above)
Comment (yunkunrao) [reply pkg/kapis/tenant/v1alpha2/register.go:329]: yeah, it's good to me

File: pkg/kapis/tenant/v1alpha2/register.go (hunk 7 above)
Comment (yunkunrao) [reply pkg/kapis/tenant/v1alpha2/register.go:338]: got it

**Pull request merged**